File: Algorithms/py/229.MajorityElementII.py
__author__ = 'qianlong'
class Solution:
    # @param nums, a list of integers
    # @return an integer
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        d = {}
        for current in nums:
            if current in d:
                d[current] += 1
            elif len(d) < 2:
                d[current] = 1
            else:
                for k in d:
                    d[k] -= 1
                d = dict([(k, v) for k, v in d.items() if v != 0])
        result = []
        for k in d:
            count = sum([1 for n in nums if n == k])
            if count * 1.0 / len(nums) > 1/3.0:
                result.append(k)
        return result

File: Algorithms/java/229.MajorityElementII.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> result = new ArrayList<>();
        Map<Integer, Integer> map = new HashMap<>();
        for (int i=0; i < nums.length; i++) {
          if (map.containsKey(nums[i])) {
            map.put(nums[i], map.get(nums[i]) + 1);
          } else if (map.size() < 2) {
            map.put(nums[i], 1);
          } else {
            for (Integer key: map.keySet()) {
              map.put(key, map.get(key) -1);
            }
            Map<Integer, Integer> new_map = new HashMap<>();
            for (Integer key: map.keySet()) {
              if (map.get(key) != 0) {
                  new_map.put(key, map.get(key));
              }
            }
            map = new_map;
          }
        }
        for (Integer key: map.keySet()) {
          int count = 0;
          for (int i=0; i < nums.length; i++) {
            if (nums[i] == key) {
              count ++;
            }
          }
          if (count * 1.0 / nums.length > 1/3.0) {
            result.add(key);
          }
        }
        return result;
    }
}
